source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveRenderContext.cs

```diff
@@ -286,24 +286,164 @@ public FontColorConfig GetForegroundColors(AdaptiveTextColor textColor)
+
+                    if (tagContent != null)
                     {
-                        elementFrameworkElement.Visibility = Visibility.Visible;
+                        elementContainers.Add(tagContent.ElementContainer);
                     }
-                    // otherwise if we read something with the format {"elementId": <id>", "isVisible": false} or we just read the id and the element is visible
-                    else if ((targetElement.IsVisible.HasValue && !targetElement.IsVisible.Value) || (!targetElement.IsVisible.HasValue && visibility == Visibility.Visible))
+                }
+            }
+
+            foreach (Grid elementContainer in elementContainers)
+            {
+                ResetSeparatorVisibilityInsideContainer(elementContainer);
+            }
+
+        }
+
+        /// <summary>
+        /// Elements are adde to the container in two ways: if the height is auto or the inserted element is a container, then the element
+        /// is added as is, if the element has height stretch, then we add an extra stack panel so it can take the remaining space
+        /// </summary>
+        /// <param name="element"></param>
+        /// <returns></returns>
+        private FrameworkElement GetRenderedElement(FrameworkElement element)
+        {
+            if (element is StackPanel containerPanel)
+            {
+                UIElement uiElement = containerPanel.Children[0];
+
+                if (uiElement is FrameworkElement frameworkElement)
+                {
+                    return frameworkElement;
+                }
+            }
+
+            return element;
+        }
+
+        private void HandleSeparatorAndSpacing(bool isFirstVisible, FrameworkElement element, TagContent tagContent)
+        {
+            // Hide the spacing / separator for the first element
+            // Separators are added as a grid
+            Grid separator = tagContent.Separator;
+
+            if (separator != null)
+            {
+                separator.Visibility = isFirstVisible ? Visibility.Collapsed : Visibility.Visible;
+            }
+            else
+            {
+                bool mustHideSpacing = (isFirstVisible && !(tagContent.SpacingHasBeenHidden));
+                bool mustShowSpacing = (!isFirstVisible && tagContent.SpacingHasBeenHidden);
+
+                if (mustHideSpacing || mustShowSpacing)
+                {
+                    FrameworkElement renderedElement = GetRenderedElement(element);
+                    var spacing = Config.GetSpacing(tagContent.Spacing);
+
+                    // The spacings are added as a margin in the top, so we have to deduct that value
+                    if (mustHideSpacing)
+                    {
+                        spacing = -spacing;
+                    }
+
+                    Thickness renderedMargin = renderedElement.Margin;
+                    renderedElement.Margin = new Thickness(renderedMargin.Left,
```

Comment:
change the spacing form being a margin to being a grid

Do this here or add a note on Bug #2816 for bleed